Stop storing FetchEvent in WebServiceWorkerFetchTaskClient

youennf opened the pull request:
#### 01a643acd37aaf87c26ee65dc46250df3eb9eb0c
<pre>
Stop storing FetchEvent in WebServiceWorkerFetchTaskClient
<a href="https://rdar.apple.com/129926640">rdar://129926640</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=275610">https://bugs.webkit.org/show_bug.cgi?id=275610</a>

Reviewed by Chris Dumez.

We do not want to extend the lifetime of FetchEvents, so we mvoe them out of WebServiceWorkerFetchTaskClient.
Instead we store the ongoing fetch tasks in ServiceWorkerGlobalScope (both WebServiceWorkerFetchTaskClient and FetchEvent).
And we handle navigation preloads from there.

Covered by existing tests.

* Source/WebCore/workers/service/ServiceWorkerGlobalScope.cpp:
(WebCore::ServiceWorkerGlobalScope::skipWaiting):
(WebCore::ServiceWorkerGlobalScope::prepareForDestruction):
(WebCore::ServiceWorkerGlobalScope::addFetchTask):
(WebCore::ServiceWorkerGlobalScope::addFetchEvent):
(WebCore::ServiceWorkerGlobalScope::removeFetchTask):
(WebCore::ServiceWorkerGlobalScope::fetchTask):
(WebCore::ServiceWorkerGlobalScope::takeFetchTask):
(WebCore::ServiceWorkerGlobalScope::hasFetchTask const):
(WebCore::ServiceWorkerGlobalScope::navigationPreloadFailed):
(WebCore::ServiceWorkerGlobalScope::navigationPreloadIsReady):
* Source/WebCore/workers/service/ServiceWorkerGlobalScope.h:
* Source/WebCore/workers/service/context/ServiceWorkerFetch.cpp:
(WebCore::ServiceWorkerFetch::dispatchFetchEvent):
* Source/WebCore/workers/service/context/ServiceWorkerFetch.h:
* Source/WebCore/workers/service/context/ServiceWorkerThread.cpp:
(WebCore::ServiceWorkerThread::queueTaskToFireFetchEvent):
* Source/WebCore/workers/service/context/ServiceWorkerThread.h:
* Source/WebCore/workers/service/context/ServiceWorkerThreadProxy.cpp:
(WebCore::ServiceWorkerThreadProxy::startFetch):
(WebCore::ServiceWorkerThreadProxy::cancelFetch):
(WebCore::ServiceWorkerThreadProxy::convertFetchToDownload):
(WebCore::ServiceWorkerThreadProxy::navigationPreloadIsReady):
(WebCore::ServiceWorkerThreadProxy::navigationPreloadFailed):
(WebCore::ServiceWorkerThreadProxy::continueDidReceiveFetchResponse):
(WebCore::ServiceWorkerThreadProxy::removeFetch):
* Source/WebCore/workers/service/context/ServiceWorkerThreadProxy.h:
* Source/WebKit/WebProcess/Storage/WebServiceWorkerFetchTaskClient.cpp:
(WebKit::WebServiceWorkerFetchTaskClient::cleanup):
(WebKit::WebServiceWorkerFetchTaskClient::setFetchEvent): Deleted.
(WebKit::WebServiceWorkerFetchTaskClient::navigationPreloadIsReady): Deleted.
(WebKit::WebServiceWorkerFetchTaskClient::navigationPreloadFailed): Deleted.
* Source/WebKit/WebProcess/Storage/WebServiceWorkerFetchTaskClient.h:

Canonical link: <a href="https://commits.webkit.org/280242@main">https://commits.webkit.org/280242@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0b78b0182de0fd833bd3393a81915c5c2240cb05

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/56053 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/35379 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/8525 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/59659 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/6489 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/58179 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/43001 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/6683 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/59659 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/4489 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/58082 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/33274 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/48341 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/59659 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/30055 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/5671 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/5493 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/52024 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/5942 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/61338 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🛠 vision ](https://ews-build.webkit.org/#/builders/visionOS-1-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/6070 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/61338 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🛠 vision-sim ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/48408 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/61338 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12420 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/31206 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/32292 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/33375 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/32039 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/workers/service/ServiceWorkerGlobalScope.cpp:
```diff
@@ -271,4 +276,85 @@ CookieStore& ServiceWorkerGlobalScope::cookieStore()
+    ASSERT(m_ongoingFetchTasks.contains(key));
+    auto iterator = m_ongoingFetchTasks.find(key);
+
+    bool isHandled = std::visit(WTF::makeVisitor([] (std::nullptr_t) {
```
Why isn't this using WTF::switchOn?

youennf replied:
Will switch back, I thought we were moving away given we are using std::variant.

szewai commented on Source/WebCore/workers/service/ServiceWorkerGlobalScope.cpp:
```diff
@@ -140,6 +142,8 @@ Page* ServiceWorkerGlobalScope::serviceWorkerPage()
 
 void ServiceWorkerGlobalScope::skipWaiting(Ref<DeferredPromise>&& promise)
 {
+    RELEASE_LOG(ServiceWorker, "ServiceWorkerThread::skipWaiting for worker %" PRIu64, thread().identifier().toUInt64());
```
This should be `ServiceWorkerGlobalScope::skipWaiting`?

youennf replied:
ok

szewai commented on Source/WebCore/workers/service/ServiceWorkerGlobalScope.cpp:
```diff
@@ -271,4 +276,85 @@ CookieStore& ServiceWorkerGlobalScope::cookieStore()
+void ServiceWorkerGlobalScope::addFetchEvent(FetchKey key, FetchEvent& event)
+{
+    ASSERT(m_ongoingFetchTasks.contains(key));
+    auto iterator = m_ongoingFetchTasks.find(key);
```
`ASSERT(iterator != m_ongoingFetchTasks.end());` to avoid double lookup?

youennf replied:
The ASSERT(contains) is more readable.

youennf committed:
Stop storing FetchEvent in WebServiceWorkerFetchTaskClient
rdar://129926640
https://bugs.webkit.org/show_bug.cgi?id=275610

Reviewed by Chris Dumez.

We do not want to extend the lifetime of FetchEvents, so we mvoe them out of WebServiceWorkerFetchTaskClient.
Instead we store the ongoing fetch tasks in ServiceWorkerGlobalScope (both WebServiceWorkerFetchTaskClient and FetchEvent).
And we handle navigation preloads from there.

Covered by existing tests.

* Source/WebCore/workers/service/ServiceWorkerGlobalScope.cpp:
(WebCore::ServiceWorkerGlobalScope::skipWaiting):
(WebCore::ServiceWorkerGlobalScope::prepareForDestruction):
(WebCore::ServiceWorkerGlobalScope::addFetchTask):
(WebCore::ServiceWorkerGlobalScope::addFetchEvent):
(WebCore::ServiceWorkerGlobalScope::removeFetchTask):
(WebCore::ServiceWorkerGlobalScope::fetchTask):
(WebCore::ServiceWorkerGlobalScope::takeFetchTask):
(WebCore::ServiceWorkerGlobalScope::hasFetchTask const):
(WebCore::ServiceWorkerGlobalScope::navigationPreloadFailed):
(WebCore::ServiceWorkerGlobalScope::navigationPreloadIsReady):
* Source/WebCore/workers/service/ServiceWorkerGlobalScope.h:
* Source/WebCore/workers/service/context/ServiceWorkerFetch.cpp:
(WebCore::ServiceWorkerFetch::dispatchFetchEvent):
* Source/WebCore/workers/service/context/ServiceWorkerFetch.h:
* Source/WebCore/workers/service/context/ServiceWorkerThread.cpp:
(WebCore::ServiceWorkerThread::queueTaskToFireFetchEvent):
* Source/WebCore/workers/service/context/ServiceWorkerThread.h:
* Source/WebCore/workers/service/context/ServiceWorkerThreadProxy.cpp:
(WebCore::ServiceWorkerThreadProxy::startFetch):
(WebCore::ServiceWorkerThreadProxy::cancelFetch):
(WebCore::ServiceWorkerThreadProxy::convertFetchToDownload):
(WebCore::ServiceWorkerThreadProxy::navigationPreloadIsReady):
(WebCore::ServiceWorkerThreadProxy::navigationPreloadFailed):
(WebCore::ServiceWorkerThreadProxy::continueDidReceiveFetchResponse):
(WebCore::ServiceWorkerThreadProxy::removeFetch):
* Source/WebCore/workers/service/context/ServiceWorkerThreadProxy.h:
* Source/WebKit/WebProcess/Storage/WebServiceWorkerFetchTaskClient.cpp:
(WebKit::WebServiceWorkerFetchTaskClient::cleanup):
(WebKit::WebServiceWorkerFetchTaskClient::setFetchEvent): Deleted.
(WebKit::WebServiceWorkerFetchTaskClient::navigationPreloadIsReady): Deleted.
(WebKit::WebServiceWorkerFetchTaskClient::navigationPreloadFailed): Deleted.
* Source/WebKit/WebProcess/Storage/WebServiceWorkerFetchTaskClient.h:

Canonical link: https://commits.webkit.org/280242@main
